File: hsafa-spaces/sdks/react-sdk/src/http.ts
import type { HsafaClientOptions } from './types.js';

export type HttpMethod = 'GET' | 'POST' | 'PATCH' | 'DELETE';

export function buildAuthHeaders(options: HsafaClientOptions): Record<string, string> {
  const headers: Record<string, string> = {};

  if (options.secretKey) {
    headers['x-secret-key'] = options.secretKey;
  } else if (options.publicKey) {
    headers['x-public-key'] = options.publicKey;
    if (options.jwt) {
      headers['Authorization'] = `Bearer ${options.jwt}`;
    }
  }

  return headers;
}

export class HttpClient {
  private baseUrl: string;
  private options: HsafaClientOptions;

  constructor(options: HsafaClientOptions) {
    this.options = options;
    this.baseUrl = options.gatewayUrl.replace(/\/+$/, '');
  }

  updateOptions(options: Partial<HsafaClientOptions>): void {
    Object.assign(this.options, options);
    if (options.gatewayUrl) {
      this.baseUrl = options.gatewayUrl.replace(/\/+$/, '');
    }
  }

  getOptions(): HsafaClientOptions {
    return { ...this.options };
  }

  private buildHeaders(): Record<string, string> {
    return {
      'Content-Type': 'application/json',
      ...buildAuthHeaders(this.options),
    };
  }

  buildUrl(path: string, params?: Record<string, string | number | undefined>): string {
    const url = new URL(`${this.baseUrl}${path}`);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    return url.toString();
  }

  async request<T>(method: HttpMethod, path: string, body?: unknown, params?: Record<string, string | number | undefined>): Promise<T> {
    const url = this.buildUrl(path, params);
    const headers = this.buildHeaders();

    const init: RequestInit = {
      method,
      headers,
    };

    if (body !== undefined && method !== 'GET') {
      init.body = JSON.stringify(body);
    }

    const response = await fetch(url, init);

    if (!response.ok) {
      let errorBody: unknown;
      try {
        errorBody = await response.json();
      } catch {
        errorBody = await response.text();
      }
      throw new HsafaApiError(response.status, errorBody);
    }

    return response.json() as Promise<T>;
  }

  async get<T>(path: string, params?: Record<string, string | number | undefined>): Promise<T> {
    return this.request<T>('GET', path, undefined, params);
  }

  async post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>('POST', path, body);
  }

  async patch<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>('PATCH', path, body);
  }

  async delete<T>(path: string): Promise<T> {
    return this.request<T>('DELETE', path);
  }

  getAuthHeaders(): Record<string, string> {
    return buildAuthHeaders(this.options);
  }

  getBaseUrl(): string {
    return this.baseUrl;
  }
}

export class HsafaApiError extends Error {
  status: number;
  body: unknown;

  constructor(status: number, body: unknown) {
    const message = typeof body === 'object' && body !== null && 'error' in body
      ? String((body as Record<string, unknown>).error)
      : `HTTP ${status}`;
    super(message);
    this.name = 'HsafaApiError';
    this.status = status;
    this.body = body;
  }
}

File: hsafa-core/old-hsafa/python-sdk/hsafa/http.py
import json
from typing import Any, Dict, Optional
from urllib.parse import urlencode

import requests

from .auth import build_auth_headers


class HsafaApiError(Exception):
    def __init__(self, status: int, body: Any):
        if isinstance(body, dict) and "error" in body:
            message = str(body["error"])
        else:
            message = f"HTTP {status}"
        super().__init__(message)
        self.status = status
        self.body = body


class HttpClient:
    def __init__(
        self,
        gateway_url: str,
        secret_key: Optional[str] = None,
        public_key: Optional[str] = None,
        jwt: Optional[str] = None,
    ):
        self.base_url = gateway_url.rstrip("/")
        self.secret_key = secret_key
        self.public_key = public_key
        self.jwt = jwt

    def get_auth_headers(self) -> Dict[str, str]:
        return build_auth_headers(self.secret_key, self.public_key, self.jwt)

    def _build_headers(self) -> Dict[str, str]:
        return {"Content-Type": "application/json", **self.get_auth_headers()}

    def build_url(self, path: str, params: Optional[Dict[str, Any]] = None) -> str:
        url = f"{self.base_url}{path}"
        if params:
            filtered = {k: str(v) for k, v in params.items() if v is not None}
            if filtered:
                url = f"{url}?{urlencode(filtered)}"
        return url

    def _request(
        self,
        method: str,
        path: str,
        body: Any = None,
        params: Optional[Dict[str, Any]] = None,
    ) -> Any:
        url = self.build_url(path, params)
        headers = self._build_headers()

        kwargs: Dict[str, Any] = {"headers": headers}
        if body is not None and method != "GET":
            kwargs["data"] = json.dumps(body)

        response = requests.request(method, url, **kwargs)

        if not response.ok:
            try:
                error_body = response.json()
            except Exception:
                error_body = response.text
            raise HsafaApiError(response.status_code, error_body)

        return response.json()

    def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Any:
        return self._request("GET", path, params=params)

    def post(self, path: str, body: Any = None) -> Any:
        return self._request("POST", path, body=body)

    def patch(self, path: str, body: Any = None) -> Any:
        return self._request("PATCH", path, body=body)

    def delete(self, path: str) -> Any:
        return self._request("DELETE", path)
